### PR TITLE
Update HcalSeverityLevelComputer to be able to use both phase0 and phase1 flags

### PR DESCRIPTION
**Pull request by jaehyeok**
This is an update on the HcalSeverityLevelComputer and the relevant codes such that the phase 1 RecHit flags can be used. 

- Needed in order to use the RecHit cleaning based on the charge asymmetry between two anodes because the current HcalSeverityLevelComputer does not support it
- Added a parameter called “phase” so that the correct set of flags (phase0 vs phase1) is chosen by era
- For back-compatibility, the current flags (phase0) are not touched
- Added an option to RecoLocalCalo/HcalRecAlgos/python/RemoveAddSevLevel.py so that warning messages can be suppressed if necessary

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
@fwyzard FYI, this will have to be updated in corresponding HLT configs for 2017 at some point

**Comment by fwyzard** [comment]
Thanks.

We will also need a 9.2.x backport, then.

**Comment by kpedro88** [comment]
@jaehyeok you should be able to submit another PR from this same branch to the CMSSW_9_2_X branch

**Comment by jaehyeok** [comment]
I just made an update to turn on the HF filters (HFPET, HFS8S1Ratio and HFLongShort) used last year. These filters will provide an additional noise reduction of around factor 4 [1] for the events with MET>100 GeV triggered by HLT_MET120 path [2]. These filters have been off so far this year. 

All the changes in this pull request affect only HF rechits, both MC and data.

Some more validation plots for the changes can be found in [3]. 

[1]  https://indico.cern.ch/event/644533/contributions/2636729/attachments/1484703/2309821/HF_Owen_files.pdf
[2] /afs/cern.ch/work/o/owen/public/trigger-2017/edm-files-june27-2017
[3]  https://indico.cern.ch/event/644533/contributions/2636729/attachments/1484703/2304779/HF_TDCcuts_DB_MET_SingleMuon.pdf

**Comment by jaehyeok** [comment]
@kpedro88 @fwyzard PR (https://github.com/cms-sw/cmssw/pull/19535) was just submitted.

**Comment by perrotta** [comment]
please test

**Comment by Martin-Grunewald** [comment]
In case HLT changes are eventually needed, please file a JIRA request
@kpedro88 @jaehyeok @souravdey1986

**Comment by jaehyeok** [comment]
JIRA ticket has been created: https://its.cern.ch/jira/browse/CMSHLT-1447.

**Comment by deguio** [comment]
@perrotta @slava77 
is this PR being reviewed?
sorry for bothering, just wanted to confirm.

**Comment by perrotta** [comment]
@deguio : I will look at it later this afternoon

**Comment by perrotta** [comment]
@jaehyeok 

In https://github.com/cms-sw/cmssw/pull/19532#issuecomment-312749957 you claim that "all the changes in this pull request affect only HF rechits, both MC and data". However, from the output of the jenkins tests there are changes also in HBHE, BeamHalo, HcalHalo, etc in all phase1 and above workflows (see e.g. https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_3_X_2017-07-03-1100+19532/21042/validateJR/all_OldVSNew_TTbar13TeV2017wf10024p0/)

Are those changes kind of expected?
Which part of the code changes did trigger them?
If they are expected, do you have some plot/presentation that summarize what expected?

**Comment by jaehyeok** [comment]
Hi @perrotta

Thanks for checking. I am taking a look.

**Comment by perrotta** [comment]
@jaehyeok : any news from your checks?

**Comment by jaehyeok** [comment]
@perrotta 

Experts are still looking. We will post a message as soon as we come to a conclusion.

**Comment by abdoulline** [comment]
@perrotta  
Hi Andrea,  this is just to keep you informed:

my tests showed you've rightly spotted the problem:
unexpected changes in "reduced" HBHE RecHits collection. This collection contains 
(1) "interesting" Hcal hits coming from ECAL isolation analysis, this part is _not_ altered by this PR
(2)  HCAL hits with SeverityLevel >0, the part which part is affected by this PR.

The problem with (2) is that an "informative" flag (bit #29) gets associated with inappropriate SeverityLevel = 8 (should stay at 0) with PR in action, thus pushing some HBHE hits to arrive in "reduced" collection... 

 @jaehyeok is kindly asked to debug further and to fix the issue.

**Comment by kpedro88** [comment]
please test

**Comment by jaehyeok** [comment]
@perrotta 

As Salavat mentioned in https://github.com/cms-sw/cmssw/pull/19532#issuecomment-315277223, the difference was caused by bit No. 29 (HBHEPulseFitBit) that had severity level=8 instead of 0.

The attached plots show that after moving HBHEPulseFitBit to severity level=0, HBHERecHitsSorted_reducedHcalRecHits_hbhereco_RECO.obj.obj.energy_ (one of the variables that Jenkins test showed discrepancy) now looks identical. The right plots show the ratio, “with PR”/“without PR”.

One minor thing found is that the severity level for HBHEIsolatedNoise was 8 instead of 5.This should not cause any differences, but I updated it for consistency with phase0.

I committed the appropriate changes (https://github.com/cms-sw/cmssw/pull/19532/commits/248d41f2993b76a191b2cb4b3fdfb10ed3871853).

thanks,
Jae

pr19532

**Comment by perrotta** [comment]
Now discrepancies remains in the DQM plots produced by jnkins, see e.g. (red is new, blue is old) the number of problematic cells in HB and HE that were all 0 before, while now that number increases:
image
image

This reflects in other DQM distributions that are quite likely only filled when a cell is "not problematic" and therefore look emptish with this PR:
image
image

All those figures are taken from 10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017 wf, but there are similar examples in all other 2017+ workflows in the DQM comparison section of the jenkins results.

Curiously enough, there are only tiny differences for HF in those comparisons, e.g.:
image


Possible explanation? Should also DQM get adapted?

**Comment by abdoulline** [comment]
@perrotta,  

your perception is correct (histos filled solely with "problematic" and only "non-problematic" cells).
CaloTowers algo has a block to mark/count "problematic" HcalRecHits  ( "bad"=9  > SeverityLevel > 0)    
https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/CaloTowersCreator/src/CaloTowersCreationAlgo.cc#L1799

But I'm confused, as "reducedHBHERecHits" collection (now OK) contains
(1) "problematic" + 
(2) "bad" hits  + 
(3) interestingECALIsolation ones (invariant number) . 

And both "good" and "reduced" (1)+(2)+(3) collections are OK now,  You observations on "problematic" cells change in CaloTowers can only mean (to me) that there some previous "bad" HBHE hits went now to "problematic" ones...
A question to Jae ( @jaehyeok ) to investigate from inside HcalSeverityLevelComputer.

**Comment by abdoulline** [comment]
I was too fast in my conclusion=rather speculation....
If "bad" hits would migrate to "problematic", good collection would increase.

 @jaehyeok
Jae, could you, please, take a closer look at all non-0 SeverityLevel assignments coming out HcalSeverityLevelComputer (w/wo this PR) from step3 of 10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017  ?

**Comment by perrotta** [comment]
Thank you Salavat.
By the way: are the changes in

**Comment by mariadalfonso** [comment]
@lathomas 
Are the HF hits used in the beamHalo Filter ? 
With this PR we are changing some severity level for the HF hits and we observed some difference in the HcalHaloData and BeamHaloSummary

**Comment by lathomas** [comment]
@mariadalfonso HF hits are not used in the method used for the halo filter decisions. 
Looking at the function producing the HCalHaloData I see that the CaloTowers are passed as an argument, but in principle, HF should be excluded [1]. 
There's a hardcoded cut: ieta <=29. I'm not sure what ieta = 29 corresponds to. Could it be the first HF tower? 


[1] https://github.com/cms-sw/cmssw/blob/master/RecoMET/METAlgorithms/src/HcalHaloAlgo.cc#L142-L156

**Comment by kpedro88** [comment]
@lathomas there is an HF tower with ieta=29, but for CaloTowers, HF towers 29 and 30 are combined to be CaloTower 30. In other words, CaloTower 29 contains only HE tower 29.

**Comment by jaehyeok** [comment]
@abdoulline 

I checked the effect of changing
RecoveredRecHitBits = cms.vstring('TimingAddedBit','TimingSubtractedBit') -> (" ").

There's no difference in the variables mentioned by @perrotta. 

Now I am taking a closer look at non-0 SeverityLevel assignments.

**Comment by abdoulline** [comment]
@jaehyeok, @igv4321,

(just for the record) 
looks like TimingAddedBit & TimingSubtractedBit (bits 17 & 16) [1]
aren't used in Phase1 (never set for MC) reco, as HcalTimingCorrector [2] is not called/invoked anymore[3]. 
These flags ("corrector") were introduced for mitigating timing sync. issues in the past. Now they become kind of obsolete...

----------------------------
[1]
https://github.com/cms-sw/cmssw/blob/09c3fce6626f70fd04223e7dacebf0b485f73f54/DataFormats/METReco/interface/HcalCaloFlagLabels.h#L49

[2]
https://github.com/cms-sw/cmssw/blob/09c3fce6626f70fd04223e7dacebf0b485f73f54/RecoLocalCalo/HcalRecAlgos/src/HcalTimingCorrector.cc
  It wasn't updated (except for minor tech. move of include) since > 8 years.

 [3]
https://cmssdt.cern.ch/lxr/search?_filestring=&_string=HcalTimingCorrector

File: RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py [hunk 1]
```diff
@@ -85,4 +86,32 @@
...
     RecoveredRecHitBits = cms.vstring('TimingAddedBit','TimingSubtractedBit'),
     DropChannelStatusBits = cms.vstring('HcalCellMask','HcalCellOff', 'HcalCellDead')
+) 
+
+# This is to use phase1 flags defined in DataFormats/METReco/interface/HcalPhase1FlagLabels.h
+from Configuration.Eras.Modifier_run2_HCAL_2017_cff import run2_HCAL_2017
+run2_HCAL_2017.toModify(hcalRecAlgos,
+    phase = cms.uint32(1),
+    SeverityLevels = {
+        0 : dict( RecHitFlags = cms.vstring('TimingFromTDC',
```
Comment (abdoulline) [inline RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py]: @jaehyeok
Jae, could you, please, remove this explicit definition (customization) in L96-L99.

Comment (jaehyeok) [inline RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py]: @abdoulline 

You are right. After removing the explicit assignments for SevLvl=0, I can confirm that the discrepant distributions become identical. I will commit the change shortly.

**Comment by perrotta** [comment]
please test

**Comment by kpedro88** [comment]
please test

**Comment by abdoulline** [comment]
Interestingly enough, the "AddOn error" link above ("Details") points to PR#19535 (92X backport of this one)  results

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19535/21595/summary.html
with some strange error messages in many hlt_MC tests:

%MSG-e JetCorrectorParametersHelper:   EventSetupRecordDataGetter:hltGetConditions@streamBeginRun  19-Jul-2017 04:22:18 CEST Run: 1 Stream: 2
Duplicate binning in record found (existing index,current index)=(0,1)
	Bins(lower bounds)=(-5.2, -9999, -9999)

**Comment by kpedro88** [comment]
@abdoulline I think that message is just a warning, not the cause of the error. It's probably spurious (couldn't connect to a file, or something).

**Comment by fwyzard** [comment]
> %MSG-e JetCorrectorParametersHelper: EventSetupRecordDataGetter:hltGetConditions@streamBeginRun 19-Jul-2017 04:22:18 CEST Run: 1 Stream: 2
> Duplicate binning in record found (existing index,current index)=(0,1)
> Bins(lower bounds)=(-5.2, -9999, -9999)

This points to an error in one f the Jet Energy Corrections used in the HLT global tag; I think it was fixed some time ago, but maybe the fix was not propagated to all GTs.

**Comment by perrotta** [comment]
With the latest fixes, there are no more visible effects on jenkins tests, also including DQM.

It has been also tested with some larger samples:
- 250 MC events in 10024.0_TTbar_2017 (the one which showed the larger discrepancies before)
- 500 real data events from 136.772_RunJetHT2016H

No effect at all also in those samples. Is it expected? I mean: shouldn't we see some effect of the turn on of the HF filters? Probably the samples I checked are not the best suited for HF... @jaehyeok : can you confirm (with a plot) that such a filter actually do something with the last version of this PR?

**Comment by trtomei** [comment]
I see the error in the "hlt_data_GRun" folder. That workflow is repacking Run2017A data and then running the HLT on top of it with auto:run2_data_GRun conditions, which translates to 92X_dataRun2_PromptLike_v5 in 9_3_X. 

The full diff between the (current) HLT GT and that GT is:
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/92X_dataRun2_HLT_v4/92X_dataRun2_PromptLike_v5

and particularly the jet corrections are not updated there:

screen shot 2017-07-20 at 15 02 25

Now, you may ask, why are the jet corrections not updated there? The answer, of course, is that this is a **PromptLike**, i.e., for offline reconstruction (and we simply didn't care to do the update there). Why are we running HLT with the offline GT? I can see from this workflow that we are running HLT and RECO in the same job now, while I thought this was not supported - has this changed?

**Comment by kpedro88** [comment]
@trtomei thanks for investigating, but since this matter is wholly unrelated to this PR, please open a separate issue for the discussion

**Comment by Martin-Grunewald** [comment]
@trtomei 
Please update the jet corrections - this addOn test tests HLT+RECO in view of future validation.
It uses the PromptLike GT.

**Comment by jaehyeok** [comment]
@perrotta 

The plots below show the eta of calo jet (pT>30 GeV) with and without this PR. I used about 2k events in /MET/Run2017B-v1/RAW that passed HLT_PFMET120_PFMHT120_IDTight [1]. The left plot shows the distributions without PR in black and with PR in red. The right plot shows the ratio of "with PR"/"without PR". 

pr19532_calojeteta

You can see that a lot jets are removed in HF while there is no difference in HBHE.  

[1] /afs/cern.ch/work/o/owen/public/trigger-2017/edm-files-june27-2017/raw-format/

**Comment by perrotta** [comment]
+1
- After some gymnastic, the update of the HcalSeverityLevelComputer adds the needed functionality without touching the default reco objects
- The effect of the HF cleaning in phase1 distributions is shown in  https://github.com/cms-sw/cmssw/pull/19532#issuecomment-316916813

**Comment by davidlange6** [comment]
hi @jaehyeok - the non-functional fill descriptions is rather suboptimal. Could you add the few other parameters needed. Thanks

**Comment by Martin-Grunewald** [comment]
Important: what's the point of adding a fillDescriptions if it is a crappy one allowing anything.
It will breack ConfDB parsing as those generated cfis have precedence over static cfi files
in ../python - and that one you keep/edit as well:  RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py.

Please either put a proper fillDescriptions (and then get rid of  RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py) or put no fillDescriptions and keep  RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py

**Comment by davidlange6** [comment]
-1

pending updated, since this breaks confdb parsing.

**Comment by jaehyeok** [comment]
@davidlange6 @Martin-Grunewald 

Apologies for the incomplete fillDescriptions.  Now It includes other parameters as well (https://github.com/jaehyeok/cmssw/commit/bd726286996bee60f8feef5b9e9eff90de448be2).

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
Please test

**Comment by perrotta** [comment]
+1
- The module configuration previously in hcalRecAlgoESProd_cfi.py is now correctly moved to the fillDescriptions() method, which also gets rid of the setAllowAnything flag.
For the rest, the same considerations as in https://github.com/cms-sw/cmssw/pull/19532#issuecomment-316943840 still hold.

**Comment by abdoulline** [comment]
@davidlange6 
David, may we get it unhold now?

**Comment by davidlange6** [comment]
merge

**Pull request merged**